shared/components/pages/take_action/take_action.component.js

```diff
@@ -96,26 +159,166 @@ class TakeActionComponent extends Panel {
     take_action.props.updateUI(ui);
   }
 
+  categoryIsTransportation(key) {
+    return this.getCategoryByAction(key) === 'transportation';
+  }
+
   isCategoryActive(category) {
-    return this.state.input_action_category === category;
+    return this.active_category_filter === category;
+  }
+
+  isStatusFilterActive(key) {
+    return this.active_status_filter === key;
   }
 
-  setActiveActionsByCategory(category) {
+  filterAndSortActions(_category, _status, sort_by) {
+    const take_action = this;
     const update = {};
-    this.actions_list.forEach((group) => {
-      if (category === group.category) {
-        update.show_actions = group.keys;
-        this.setState(update);
+    const category = _category || this.active_category_filter;
+    const status = _status || this.active_status_filter;
+    let updated_sort_by;
+
+    if (category === 'all') {
+      update.show_actions = take_action.all_actions;
+    } else {
+      update.show_actions = take_action.getActionsByCategory(category);
+    }
+
+    if (status !== 'all') {
+      update.show_actions = take_action.getActionsByStatus(update.show_actions, status);
+    }
+
+    if (!sort_by) {
+      updated_sort_by = 'inactive';
+    } else if (sort_by !== 'inactive') {
+      update.show_actions = take_action.sortActions(update.show_actions, sort_by);
+      updated_sort_by = sort_by;
+    }
+
+    take_action.props.updateUI({ id: 'take_action', data: { sort_by: updated_sort_by } });
+    take_action.setState(update);
+  }
+
+  setCategory(category) {
+    const take_action = this;
+    take_action.props.updateUI({ id: 'take_action', data: { category_filter: category } });
+    take_action.filterAndSortActions(category, '', '');
+  }
+
+  getActionsByCategory(category) {
+    let actions;
+    this.actions_by_category.forEach((group) => {
+      if (category === group.id) {
+        actions = group.keys;
       }
     });
+    return actions;
   }
 
-  setCategory(category) {
+  getActionsByStatus(actions_to_filter, status) {
     const take_action = this;
-    take_action.setState({
-      input_action_category: category,
+    let filtered;
+
+    if (status === 'pledged' || status === 'completed') {
+      filtered = Object.keys(take_action.actions_profile.get(status).toJS());
+      filtered = actions_to_filter.filter(key => filtered.includes(key));
+    } else if (status === 'not_relevant') {
+      filtered = take_action.actions_profile.get(status).toJS();
+      filtered = actions_to_filter.filter(key => filtered.includes(key));
+    } else if (status === 'not_pledged') {
+      const pledged = Object.keys(take_action.actions_profile.get('pledged').toJS());
+      const not_pledged = actions_to_filter.filter(key => !pledged.includes(key));
+
```

Comment:
Avoid casting to JS when you can - it makes easier reading and less time figuring out whether we are dealing with Immutable or plain JS: 

https://facebook.github.io/immutable-js/docs/#/Map/filter